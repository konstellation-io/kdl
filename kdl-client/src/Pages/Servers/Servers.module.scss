@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.container {
  position: relative;
  height: 100%;
  width: 100%;
}

.content {
  position: relative;
  padding: 3 * $grid-unit 11 * $grid-unit;
  height: 100%;
  width: 100%;
  background-color: palette(base, 800);
  overflow: auto;
}
