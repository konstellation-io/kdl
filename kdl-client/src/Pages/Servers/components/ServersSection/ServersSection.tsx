import AddServer from '../Server/AddServer';
import { Button } from 'kwc';
import IconAdd from '@material-ui/icons/Add';
import React from 'react';
import { Server } from 'Hooks/useServers';
import styles from './ServersSection.module.scss';

type Props = {
  title: string;
  subtitle: string;
  servers: Server[];
  ServerComponent: React.FC<any>;
  addServerRoute?: string;
};
function ServersSection({
  title,
  subtitle,
  servers,
  ServerComponent,
  addServerRoute,
}: Props) {
  return (
    <div className={styles.container}>
      <div className={styles.titleSection}>
        <p className={styles.title}>{title}</p>
        {addServerRoute && (
          <Button
            label="ADD SERVER"
            Icon={IconAdd}
            className={styles.action}
            to={addServerRoute}
          />
        )}
      </div>
      <p className={styles.subtitle}>{subtitle}</p>
      <div className={styles.servers}>
        {servers.map((server) => (
          <ServerComponent {...server} key={server.id} serverId={server.id} />
        ))}
        {addServerRoute && <AddServer label="ADD SERVER" to={addServerRoute} />}
      </div>
    </div>
  );
}

export default ServersSection;
