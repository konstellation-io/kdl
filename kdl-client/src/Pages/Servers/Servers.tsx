import { LocalServer, RemoteServer } from './components/Server/Server';
import React, { useEffect, useState } from 'react';
import useServers, { Server, ServerType } from 'Hooks/useServers';

import ROUTE from 'Constants/routes';
import ServersBar from './components/ServersBar/ServersBar';
import ServersSection from './components/ServersSection/ServersSection';
import styles from './Servers.module.scss';
import { useForm } from 'react-hook-form';

type FormData = {
  serverSearch: string;
};

function Servers() {
  const { servers } = useServers();
  const [filteredServers, setFilteredServers] = useState<Server[]>(servers);

  const { setValue, unregister, register, watch } = useForm<FormData>({
    defaultValues: { serverSearch: '' },
  });

  useEffect(() => {
    register('serverSearch');
    return () => unregister('serverSearch');
  }, [register, unregister, setValue]);

  const serverSearch = watch('serverSearch');
  useEffect(() => {
    setFilteredServers(
      servers.filter(
        (server) =>
          server.type === ServerType.LOCAL || server.url?.includes(serverSearch)
      )
    );
  }, [servers, serverSearch]);

  const localServers: Server[] = servers.filter(
    (server) => server.type === ServerType.LOCAL
  );
  const filteredRemoteServers = filteredServers.filter(
    (server) => server.type === ServerType.REMOTE
  );
  const nServers = localServers.length + filteredRemoteServers.length;

  return (
    <div className={styles.container}>
      <ServersBar setValue={setValue} nServers={nServers} />
      <div className={styles.content}>
        <ServersSection
          title="LOCAL SERVER"
          subtitle="ONLY ONE LOCAL SERVER IS AVAILABLE IN KONSTELLATION"
          servers={localServers}
          ServerComponent={LocalServer}
          addServerRoute={
            localServers.length === 0
              ? ROUTE.CHECK_LOCAL_SERVER_REQUIREMENTS
              : undefined
          }
        />
        <ServersSection
          title="REMOTE SERVERS"
          subtitle="YOU MAY HAVE AS MANY SERVERS AS YOU WANT"
          servers={filteredRemoteServers}
          ServerComponent={RemoteServer}
          addServerRoute={ROUTE.CONNECT_TO_REMOTE_SERVER}
        />
      </div>
    </div>
  );
}

export default Servers;
