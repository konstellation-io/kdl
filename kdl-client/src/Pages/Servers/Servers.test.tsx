import React from 'react';
import Servers from './Servers';
import ServersSection from './components/ServersSection/ServersSection';
import { shallow } from 'enzyme';

jest.mock('Hooks/useServers', () =>
  jest.fn(() => ({
    servers: [{ type: 'local' }, { type: 'remote' }, { type: 'remote' }],
  }))
);
jest.mock('Hooks/useServers', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    servers: [{ type: 'local' }, { type: 'remote' }, { type: 'remote' }],
  })),
  ServerType: { LOCAL: 'local', REMOTE: 'remote' },
}));

const component = shallow(<Servers serverSearch="someSearch" />);

describe('Servers component', () => {
  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right amount of servers', () => {
    expect(component.find(ServersSection).first().props().servers.length).toBe(
      1
    );
    expect(component.find(ServersSection).last().props().servers.length).toBe(
      2
    );
  });
});
