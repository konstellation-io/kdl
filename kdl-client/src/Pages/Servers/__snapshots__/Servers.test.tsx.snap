// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Servers component Component match snapshot 1`] = `
<div
  className="container"
>
  <ServersBar
    nServers={3}
    setValue={[Function]}
  />
  <div
    className="content"
  >
    <ServersSection
      ServerComponent={[Function]}
      servers={
        Array [
          Object {
            "type": "local",
          },
        ]
      }
      subtitle="ONLY ONE LOCAL SERVER IS AVAILABLE IN KONSTELLATION"
      title="LOCAL SERVER"
    />
    <ServersSection
      ServerComponent={[Function]}
      addServerRoute="/connect-to-remote-server"
      servers={
        Array [
          Object {
            "type": "remote",
          },
          Object {
            "type": "remote",
          },
        ]
      }
      subtitle="YOU MAY HAVE AS MANY SERVERS AS YOU WANT"
      title="REMOTE SERVERS"
    />
  </div>
</div>
`;
